lightning/backend/session.go

```diff
@@ -38,11 +45,184 @@ func (*invalidIterator) Valid() bool {
+	i := &memDbIter{
+		iter:    m.sandbox.NewIterator(),
+		start:   k,
+		end:     upperBound,
+		reverse: false,
+	}
+
+	if k == nil {
+		i.iter.SeekToFirst()
+	} else {
+		i.iter.Seek(k)
+	}
+	return i, nil
+}
+
+func (m *memDbBuffer) IterReverse(k kv.Key) (kv.Iterator, error) {
+	return nil, nil
+}
+
+// Get returns the value associated with key.
+func (m *memDbBuffer) Get(ctx context.Context, k kv.Key) ([]byte, error) {
+	return nil, nil
+}
+
+// Set associates key with value.
+func (m *memDbBuffer) Set(k kv.Key, v []byte) error {
+	if len(v) == 0 {
+		return errors.Trace(kv.ErrCannotSetNilValue)
+	}
+	if len(k)+len(v) > m.entrySizeLimit {
+		return kv.ErrEntryTooLarge.GenWithStackByArgs(m.entrySizeLimit, len(k)+len(v))
+	}
+
+	m.sandbox.Put(k, v)
+	if m.Size() > int(m.bufferSizeLimit) {
+		return kv.ErrTxnTooLarge.GenWithStackByArgs(atomic.LoadUint64(&kv.TxnTotalSizeLimit))
+	}
+	return nil
+}
+
+// Delete removes the entry from buffer with provided key.
+func (m *memDbBuffer) Delete(k kv.Key) error {
+	return nil
+}
+
+// Size returns sum of keys and values length.
+func (m *memDbBuffer) Size() int {
+	return m.sandbox.Size()
+}
+
+// Len returns the number of entries in the DB.
+func (m *memDbBuffer) Len() int {
+	return m.sandbox.Len()
+}
+
+func (m *memDbBuffer) NewStagingBuffer() kv.MemBuffer {
+	return &memDbBuffer{
+		sandbox:         m.sandbox.Derive(),
+		entrySizeLimit:  kv.TxnEntrySizeLimit,
+		bufferSizeLimit: m.bufferSizeLimit - uint64(m.sandbox.Size()),
+	}
+}
+
+func (m *memDbBuffer) Flush() (int, error) {
+	// There is no need to check size limit,
+	// because the size limit is maintain when derive this buffer.
+	return m.sandbox.Flush(), nil
+}
+
+// Discard must do nothing because it will discard every time when AddRecord finished
+// It's hard code in TiDB after this PR(https://github.com/pingcap/tidb/pull/15931/files) merged.
+func (m *memDbBuffer) Discard() {
+	// Do nothing
+}
+
+// RealDiscard will truly discard all staging kvs, after lightning translate all kvs into kvpairs
+// It will trigger manually.
+func (m *memDbBuffer) RealDiscard() {
+	m.sandbox.Discard()
+}
+
+// WalkMemBuffer iterates all buffered kv pairs in memBuf
+func WalkMemBuffer(memBuf *memDbBuffer, f func(k kv.Key, v []byte) error) error {
+	iter, err := memBuf.Iter(nil, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	defer iter.Close()
+	for iter.Valid() {
+		if err = f(iter.Key(), iter.Value()); err != nil {
+			return errors.Trace(err)
+		}
+		err = iter.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
 // transaction is a trimmed down Transaction type which only supports adding a
 // new KV pair.
 type transaction struct {
 	kv.Transaction
-	kvPairs []common.KvPair
+	kvPairs     []common.KvPair
+	bufferStore *memDbBuffer
+}
+
+func (t *transaction) NewStagingBuffer() kv.MemBuffer {
+	if t.bufferStore == nil {
+		t.bufferStore = NewMemDbBuffer()
+	}
+	return t.bufferStore
```

Comment:
why can't we use `t` as the MemBuffer?

Reply:
because of [this](https://github.com/pingcap/tidb-lightning/pull/324/files#diff-ea68acd8c8c86c13a06db5cb0b8bd058R180-R190) , current tidb will discard staging kv before `AddRecord` return, so our `memDbBuffer` is different from `kv.MemBuffer`,  it has `RealDiscard` method to do real discard.

Reply:
we could implement `(*transaction).Discard()` to do nothing?